### ISSUE TITLE
False positive for `-Walign-mismatch` with C `aligned` attribute in Clang

### ISSUE THREAD

**Issue by Deewiant** [llvm/llvm-project#57564]
Example code:

```c
typedef char char_aligned2 __attribute__((aligned(2)));
int foo(char_aligned2 *p) {
   return *p + 1;
}
int bar(void) {
   char x[] __attribute__((aligned(2))) = "foo";
   return foo(x);
}
```

Result:

```
$ clang asdf.c -Weverything -Wno-missing-prototypes -S -emit-llvm
asdf.c:7:15: warning: passing 1-byte aligned argument to 2-byte aligned parameter 1 of 'foo' may result in an unaligned pointer access [-Walign-mismatch]
   return foo(x);
              ^
1 warning generated.
```

The emitted code seems to do what I expect, in that `foo` does a `load` with `align 2` and `bar` does an `alloca` with `align 2`, so it all seems to work out correctly in the end but the diagnostic doesn't match the codegen: it's not a "1-byte aligned argument".

A `(void*)` cast is a workaround but of course effectively disables the diagnostic.

Tested with 14.0.6:

```
$ clang --version
clang version 14.0.6
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

**Comment by mizvekov** [comment]
```
char x[] __attribute__((aligned(2))) = "foo";
```
The problem here is that the attribute is applied to the declaration.
The array will be aligned, however it's type will have no such information.

If you declare it like this, it should work:
```
char_aligned2 x[] = "foo";
```
The alignment is attached to the TypedefName, but that however is fine since that is part of the type now.

To support your purported use case, we would have to implement specifying aligned attribute on the type directly, which we don't. It can presently be attached only to Variables and TypedefNames.

**Comment by Deewiant** [comment]
Oh, I didn't realize that works in Clang, because unfortunately GCC (12.2.0) in turn doesn't like that formulation:

```
$ gcc -c asdf.c
asdf.c: In function ‘bar’:
asdf.c:6:4: error: alignment of array elements is greater than element size
    6 |    char_aligned2 x[] = "foo";
      |    ^~~~~~~~~~~~~
```

While the complaint itself seems to just be an internal GCC limitation, I somewhat rationalized it by imagining that this declaration would mean that each element in the array should be 2-byte aligned instead of just the first.
